debian/patches/backport-ubuntu-pro-rename.patch

```diff
@@ -0,0 +1,4187 @@
++            match=re.escape("Failure enabling Ubuntu Pro service(s): esm, cc"),
++        ):
++            configure_pro(
++                token="SomeToken", enable=["esm", "cc", "fips", "asdf"]
++            )
++        assert m_subp.call_args_list == [
++            mock.call(
++                ["pro", "attach", "--no-auto-enable", "SomeToken"],
++                logstring=["pro", "attach", "--no-auto-enable", "REDACTED"],
++                rcs={0, 2},
++            ),
++            mock.call(
++                [
++                    "pro",
++                    "enable",
++                    "--assume-yes",
++                    "--format",
++                    "json",
++                    "esm",
++                    "cc",
++                    "fips",
++                    "asdf",
++                ],
++                capture=True,
++                rcs={0, 1},
++            ),
++        ]
++        assert (
++            MPATH,
++            logging.WARNING,
++            "Failure enabling `esm`: Invalid esm credentials",
++        ) in caplog.record_tuples
++        assert (
++            MPATH,
++            logging.WARNING,
++            "Failure enabling `cc`: Invalid cc credentials",
++        ) in caplog.record_tuples
++        assert (
++            MPATH,
++            logging.WARNING,
++            "Failure of type `system`: Cannot enable unknown service 'asdf'",
++        ) in caplog.record_tuples
++        assert 'Failure enabling "fips"' not in caplog.text
++
++    def test_pro_enable_unexpected_error_codes(self, m_subp):
++        def fake_subp(cmd, capture=None, **kwargs):
++            if cmd[:2] == ["pro", "enable"] and capture:
++                raise subp.ProcessExecutionError(exit_code=255)
++            return subp.SubpResult(json.dumps({"errors": []}), "")
++
++        m_subp.side_effect = fake_subp
++
++        with pytest.raises(
++            RuntimeError,
++            match=re.escape("Error while enabling service(s): esm"),
++        ):
++            configure_pro(token="SomeToken", enable=["esm"])
++
++    def test_pro_enable_non_json_response(self, m_subp):
++        def fake_subp(cmd, capture=None, **kwargs):
++            if cmd[:2] == ["pro", "enable"] and capture:
++                return subp.SubpResult("I dream to be a Json", "")
++            return subp.SubpResult(json.dumps({"errors": []}), "")
++
++        m_subp.side_effect = fake_subp
++
++        with pytest.raises(
++            RuntimeError,
++            match=re.escape("Pro response was not json: I dream to be a Json"),
++        ):
++            configure_pro(token="SomeToken", enable=["esm"])
++
++
++JSONSCHEMA_SKIP_REASON = (
++    "deprecation unraised as jsonschema ver can't merge $defs and inline keys"
++)
++
++
++class TestUbuntuProSchema:
++    @pytest.mark.parametrize(
++        "config, expectation, skip_reason",
++        [
++            pytest.param({"ubuntu_pro": {}}, does_not_raise(), ""),
++            pytest.param(
```

Comment:
I'm confused by this test case. The end result of cloud-init not warning on 'ubuntu_advantage' keys works like how I'd expect, but why are we checking jsonschema version in order to skip?

If I try a test case below this one, e.g., `{"ubuntu_pro": {"token": "win", "invalidkey": ""}}`, I get the error I would expect, but if I do `ubuntu_advantage: {}`, I get no error. How is jsonschema version related to whether this works or not?

Reply:
I believe this was because older json schema didn't handle merging [inline $defs with other keys at the same level in the dict path](https://github.com/canonical/cloud-init/blob/main/tests/unittests/config/test_cc_ubuntu_pro.py#L432).The result was that the merged dict was more permissive of extra keys on older versions of jsonschema than we'd ideally want, I think we opted of avoid resolution of that on LTS releases because schema validation was a new feature that didn't necessarily have to be as strict on older releases.